File: 1121-partition-array-for-maximum-sum/1121-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        if arr is None or len(arr) == 0:
            return 0
        
        size = len(arr)

        dp = [0 for i in range(size)]
        dp[0] = arr[0]

        for i in range(1, size):
            maximum = arr[i]
            for j in range(1, k + 1):
                if i - j + 1 >= 0:
                    maximum = max(maximum, arr[i - j + 1])
                    if i - j >= 0:
                        dp[i] = max(dp[i], maximum * j + dp[i - j])
                    else:
                        dp[i] = max(dp[i], maximum * j)
                j -= 1
        
        return dp[-1]

File: 1121-partition-array-for-maximum-sum/1121-partition-array-for-maximum-sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        if(arr ==null || arr.length == 0) return 0;

        int size = arr.length;
        int[] dp = new int[size];
        dp[0] = arr[0];

        for(int i = 1; i < size; i++){
            int max = arr[i];
            for(int j = 1; j <= k; j++){
                if(i - j + 1 >= 0){
                    if(max < arr[i - j + 1]){
                        max = arr[i - j + 1];
                    }
                    if(i - j >= 0){
                        dp[i] = Math.max(dp[i], max * j + dp[i - j]);
                    }
                    else{
                        dp[i] = Math.max(dp[i], max * j);
                    }
                }
            }
        }
        return dp[dp.length - 1];
    }
}